proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java

```diff
@@ -223,41 +221,33 @@ private void initializePlayer(GameProfile profile, boolean onlineMode) {
           mcConnection,
           inbound.getVirtualHost().orElse(null));
       this.connectedPlayer = player;
-
-      if (!server.registerConnection(player)) {
+      if (!server.canRegisterConnection(player)) {
         player.disconnect(VelocityMessages.ALREADY_CONNECTED);
         return CompletableFuture.completedFuture(null);
       }
 
+      // Fire DisconnectEvent if player will be disconnected/kicked during login and next events. 
+      // This will protect us from bad plugins, those will store Player object from events
+      // and remove it on disconnect
+      mcConnection.setSessionHandler(new InitialConnectSessionHandler(player));
```

Comment:
This isn't required, since `LoginSessionHandler` will tear down the player. Look at the implementation of `disconnected()`.

Reply:
Oh. did not see it